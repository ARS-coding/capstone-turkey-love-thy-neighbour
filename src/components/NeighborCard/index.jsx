import React from "react";

import { Card } from "react-bootstrap";

import { useDispatch, useSelector } from "react-redux";

import { useTranslation } from "react-i18next";

import { send } from "emailjs-com";

import { v4 as uuidv4 } from "uuid";
import { NeighborCardButton } from "../CustomButtons";

import firebaseApp, { firestore } from "../../firebaseConfig";

import "./index.scss";

function NeighborCard({
  photo,
  firstName,
  lastName,
  age,
  gender,
  email,
  setSelectedNeighbor,
  senderEmail,
  senderFullName,
  setEmailAlertStatus,
}) {
  const dispatch = useDispatch();
  const uid = useSelector((state) => state.user.authCred?.uid);

  const { t } = useTranslation();

  function sendEmail() {
    setEmailAlertStatus("empty");
    send("service_9rwjsp6", "template_qlu5ttf", {
      from_name: senderFullName,
      from_email: senderEmail,
      to_name: `${firstName} ${lastName}`,
      to_email: email,
    })
      .then(() => setEmailAlertStatus("success"))
      .catch(() => setEmailAlertStatus("danger"));
  }

  function handleInvitation() {
    sendEmail();
    firestore // create invitation notification for the invited user
      .collection("users")
      .where("email", "==", email)
      .get()
      .then((querySnapshot) => {
        const { docs } = querySnapshot;
        const firstDoc = docs[0];
        const firstDocData = firstDoc.data();
        firstDocData.invitationNotifications.push({
          message: `How was your meeting with ${senderFullName}?`,
          id: uuidv4(),
        });
        firstDoc.ref.update(firstDocData);
      });
    firestore // create invitation for the inviter user
      .collection("users")
      .doc(uid)
      .update({
        // eslint-disable-next-line import/no-named-as-default-member
        invitationNotifications: firebaseApp.firestore.FieldValue.arrayUnion({
          message: `How was your meeting with ${firstName} ${lastName}?`,
          id: uuidv4(),
        }),
      });
  }

  return (
    <Card className="neighbor-card mb-2 mx-auto">
      <Card.Img
        variant="top"
        src={photo}
        className="img-fluid"
        id={email}
        onClick={(event) => {
          firestore
            .collection("users")
            .where("email", "==", event.target.id)
            .get()
            .then((snapshot) => setSelectedNeighbor(snapshot.docs[0].data()));
          dispatch({ type: "neighborSummary" });
        }}
      />
      <Card.Body className="d-flex flex-column justify-content-center">
        <h2 className="text-center">
          {firstName} <br /> {lastName}
        </h2>
        <small className="text-center">{`${age} / ${gender}`}</small>
        <NeighborCardButton onClick={handleInvitation}>
          {t("neighborcards_button")}
        </NeighborCardButton>
      </Card.Body>
    </Card>
  );
}

export default NeighborCard;
