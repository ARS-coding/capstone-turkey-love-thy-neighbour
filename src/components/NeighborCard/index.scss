.neighbor-card {
  max-width: 250px;
  height: 320px;

  @media screen and (max-width: 991px) {
    width: 230px;
    margin: 0 auto;
  }

  & > img {
    width: 100%;
    min-height: 180px;
    max-height: 180px;
    padding: 10px 10px 0 10px;
  }

  & > div.card-body {
    padding: 10px;
  }
}
