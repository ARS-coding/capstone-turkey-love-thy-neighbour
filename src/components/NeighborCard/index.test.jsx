import React from "react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import NeighborCard from ".";

describe("NeighborCard", () => {
  const props = {
    photo: "test",
    firstName: "test",
    lastName: "test",
    age: "test",
    gender: "test",
    email: "test",
    setSelectedNeighbor: "test",
    senderEmail: "test",
    senderFullName: "test",
    setEmailAlertStatus: "test",
  };
  const mockState = {
    user: {
      authCred: {
        uid: "Test Uid",
      },
    },
  };
  const mockStore = configureStore();
  let store;
  let tree;

  beforeEach(() => {
    store = mockStore(mockState);
    tree = renderer
      .create(
        <Provider store={store}>
          <NeighborCard {...props} />
        </Provider>
      )
      .toJSON();
  });

  it("renders NeighborCard correctly", () => {
    expect(tree).toMatchSnapshot();
  });
});
