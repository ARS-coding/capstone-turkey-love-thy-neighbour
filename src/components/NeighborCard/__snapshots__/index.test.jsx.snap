// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeighborCard renders NeighborCard correctly 1`] = `
<div
  className="neighbor-card mb-2 mx-auto card"
>
  <img
    className="card-img-top img-fluid"
    onClick={[Function]}
  />
  <div
    className="d-flex flex-column justify-content-center card-body"
  >
    <div
      className="text-center card-title h5"
    >
      undefined undefined
    </div>
    <small
      className="text-center"
    >
      undefined / undefined
    </small>
    <button
      className="sc-bdnxRM sc-fujyAs fhyKDU PaHUc"
      onClick={[Function]}
    >
      Invite To Meet!
    </button>
  </div>
</div>
`;
